README
joystick.txt

ChangeLog
		Fixed throttle automagic calibration for Logitech WMED 3D
		Fixed computing the number of 10-bit axes in joy-logitech
		Fixed stream-based autotiming
		Extended the timeout for joy-assassin, there seem to be timing glitches in the protocol
		Added debug modes to joy-logitech and joy-sidewinder	
		Added Firmware bug reporting to joy-logitech
		Fixed joy-pci CHF mode, doc, and parameter parsing
		Fixed __setup stuff
		Fixed Makefile, Config.in everywhere
		Added speed measurement to sidewinder driver
		Added automatic gameport speed adjustment to logitech, gravis, thrustmaster, assassin and creative drivers
		Dropped generic gameport speed measurement
		Dropped kernel v2.0 support
		Added several workarounds for Logitech ADI/USB device firmware bugs
		Fixed analog joystick FCS/PXY interaction
		Made a 4-axis, 4-button joystick the default instead of two 2-axis 2-button joysticks	
		Added a fix for the latest USB/GamePort Logitech joysticks
		Rewrote joy-pci to shorten it quite a bit, if it doesn't work, it's my fault
		Fixed PSX doc
		Added joy-pci module by Raymond Ingles (support for Trident 4DWave and Aureal Vortex)
		Changed asm/spinlock.h to linux/spinlock.h for 2.3
		Fixed badly handled kmalloc failure in js_open, bug found by Hartwig Mebold
		Fixed N64 axis/button counts
		Fixed axis calibration for the Fusion pad and all unknown Thrustmaster joysticks
		Fixed 0.8.0.2 2nd joystick lacking power
		Fixed parameter array lengths in joy-analog, joy-lightning and joy-assassin
		Added 0.8.0.2 second joystick support to joy-db9.c
		Joystick will now need CONFIG_M686 on v2.0 kernels to enable RDTSC - because of Cyrix CPUs
		Updated PSX to David Kuder's latest version
		Updated N64 to John Dahlstrom's latest version
		Added kernel 2.3 __setup statements
		Added SpaceBall 4000 FLX driver by Joseph Krahn
		Added O_NONBLOCK to jsattach to make it work with newer kernels
		Added support for the Thrustmaster Fusion GamePad
		Added conf.modules info to joystick.txt
		Fixed negative value calibration bug in jscal (fix by John Dahlstrom)
		Added manpages for the utilities by Dan Gohman
		Extended checksummed area for SideWinder FFW
		Added some workarounds and tunings for SW PP into joy-sidewinder.c
		Updated to 2.2.11 and 2.3.13
		Fixed jscal to report kernel/jscal versions if they differ
		Included Shirakura Tetsuya's fix for the Genesis 6button pad
		Switched to hweight8 bit counting routine in joy-analog.h, thanks to Thomas Sailer
		Added a 2.3 kernel patch
		Added John Dahlstrom's N64 controller support via DirectPadPro 5.0 adapters
		Included Tim Waugh's fix for parport in 2.3 kernels
		Adapted to wait queue changes in 2.3 kernels
		Added a missing else to the FFP detection
		Removed cli()'s from PSX read routine
		Fixed joy-warrior usecount bug
		Don't require bidirectional gameport for 0.8.0.2 type joysticks
		Fixed parport search bug in joy-db9.c

Version 1.2.14 released Mar 08 1999

		Fixed Pentium (TSC) compiletime detection for 2.0 kernels
		Fixed number of buttons fro TM Millenium 3D Inceptor
		Don't believe gameport until write (thrustmaster, analog) - trouble on Intel MBs
		Added check for 1-bit mode in 3-bit mode on SW 3d Pro
		Added SW 3d Pro switch as a button
		Added SW 3d Pro ID packet contents printing
		More robust (against ISA DMA) calibration for joystick.c and joy-analog.c
		Added packet fixing for 3d Pro without optimization - helps on hostile soundcards
		SW 3d Pro optimization is disabled in case of many bit errors
		Made sidewinder detection more immune to bit misses
		Removed CRTSCTS from warrior serial line config
		Fixed select() behavior with 0.x interface on 2.1 kernels 
		Fixed select() behavior with init events on 2.0 kernels 
		Fixed axis numbering bug in joy-amiga
		Added workaround for SW Force Feedback Pro ID quirk
		Fixed in-kernel turbografx driver forgotten initialization
		Documentation updates
		Added workaround for SW FreeStyle Pro ID quirk
		joy-sidewinder is now always using 3-bit mode when possible
		Reworked SideWinder identification 
		Fixed button problems on Logitech pads 
		Rewrote joy-analog driver timing routine to use an ultra-stable dual-timer method
		Fixed long standing (invisible) RDTSC bug
		Added SideWinder Force Feedback Wheel support by Benoit Triquet
		Radically reduced generic timing code
		Extended joy-thrustmaster to handle any ThrustMaster BSP device
		Moved to one line discipline per serial joystick type, as a temporary solution
		Fixed oops when jsattach dies before /dev/jsX is closed
		Kernel now exports joystick symbols, fixing the 'inkernel mainmodule, modular submodules' bug
		Fixed Assasin -> Assassin spelling
		Fixed Magellan initial button state
		Reimplemented SpaceTec SpaceOrb 360/Spaceball Avenger driver
		Added Logitech WingMan Warrior driver - joy-warrior.c
		Fixes in the Magellan driver
		Added Creative Blaster GamePad driver - joy-creative.c
		Added Sidewinder FreeStyle Pro driver by Justin Wake
		Slightly improved Genesis6 timing sequence
		Enlarged JS_PSX_DELAY to 15 for better compatibility
		Added Magellan/SpaceMouse driver - joy-magellan.c
		Made Logitech packet read routine fail on invalid bit change
		Fixed -p and -s parameters in jscal, and cleaned up the program a bit
		Rewritten Logitech ADI driver so that it supports any Logitech ADI device
		Added SpaceTec SpaceOrb 360 driver by David Thompson - joy-spaceorb.c

Version 1.2.13 released Dec 04 1998

		More doc updates
		Added jscal and jstest to make install
		Updated Configure.help and Config.in
		Adopted kernel 2.1.128 Configure.help reformats and spellfixes
		Changed 'if (ptr)' to 'if (ptr!=NULL)' hopefully everywhere
		Fixed ThrustMaster joystick support
		Changed js_[un]register_{port|device} so that kmalloc / kfree are called only once
		Added checking for kmalloc return value in js_register_{port|device}
		Added more locking in js_open()
		Added checking for kmalloc return value in js_open() in joystick.c

joydump.c
/*
 *  joydump.c  Version 1.2
 *
 *  Copyright (c) 1996-1999 Vojtech Pavlik
 *
 *  Sponsored by SuSE
 */

/*
 * This is just a very simple driver that can dump the data
 * out of the joystick port into the syslog ...
 */

/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 * 
 * Should you need to contact me, the author, you can do so either by
 * e-mail - mail your message to <vojtech@suse.cz>, or by paper mail:
 * Vojtech Pavlik, Ucitelska 1576, Prague 8, 182 00 Czech Republic
 */

#include <asm/io.h>
#include <asm/system.h>
#include <linux/config.h>
#include <linux/module.h>
#include <linux/ioport.h>
#include <linux/joystick.h>
#include <linux/kernel.h>
#include <linux/delay.h>

MODULE_AUTHOR("Vojtech Pavlik <vojtech@suse.cz>");

#define JS_PORT 0x201

#ifdef __i386__
#ifdef CONFIG_X86_TSC
#define GET_TIME(x)	__asm__ __volatile__ ( "rdtsc" : "=a" (x) : : "dx" )
#endif
#elif __alpha__
#define GET_TIME(x)	__asm__ __volatile__ ( "rpcc %0" : "=r" (x) )
#endif

#ifndef GET_TIME
#warning "Using unreliable time source."
#define FAKE_TIME
static unsigned long dumptime = 0;
#define GET_TIME(x)	do { x = dumptime++; } while(0)
#endif

#define BUF_SIZE 256

struct datatype {
	unsigned int time;
	unsigned char data;
};

static struct datatype buf[BUF_SIZE];

int init_module(void)
{
	int i, j;
	unsigned int t, t1;
	int timeout, div;
	unsigned long flags;
	unsigned char u;

	printk(KERN_INFO "joydump: ,------------------- START ------------------.\n");

/*
 * Calibrate.
 */

	__save_flags(flags);
	__cli();
	GET_TIME(t);
	udelay(1000);
#ifdef FAKE_TIME
	dumptime += 1000;
#endif
	GET_TIME(t1);
	__restore_flags(flags);

	timeout = t1 - t;
	div = timeout / 1000;

	if (div < 1) {
		printk(KERN_INFO "joydump: |             CLOCK NOT TICKING!             |\n");
		printk(KERN_INFO "joydump: `-------------------- END -------------------'\n");
	}

	printk(KERN_INFO "joydump: | Timer ticks at %3d.%02d MHz.                 |\n",
		timeout / 1000, timeout / 10 % 100);

	__save_flags(flags);
	__cli();
	GET_TIME(t);
	inb(JS_PORT); inb(JS_PORT); inb(JS_PORT); inb(JS_PORT); inb(JS_PORT);
	inb(JS_PORT); inb(JS_PORT); inb(JS_PORT); inb(JS_PORT); inb(JS_PORT);
	GET_TIME(t1);
	__restore_flags(flags);

	printk(KERN_INFO "joydump: | I/O at %#x takes %d.%02d us.                |\n",
		JS_PORT, (t1 - t) / div / 10, (t1 - t) * 10 / div % 100);

	timeout *= 40;

/*
 * Gather data.
 */

	__save_flags(flags);
	__cli();

	u = inb(JS_PORT);
	GET_TIME(t1);
	t = t1;
	buf[0].data = u;
	buf[0].time = t1;
	i = 1;

	outb(0xff,JS_PORT);

	while (i < BUF_SIZE && t1 - t < timeout) {

		GET_TIME(t1);
		buf[i].data = inb(JS_PORT);

		if (buf[i].data != u) {
			u = buf[i].data;
			buf[i].time = t1;
			i++;
		}
	}

	__restore_flags(flags);

/*
 * Dump data.
 */

	t = i;

	printk(KERN_INFO "joydump: >------------------- DATA -------------------<\n");
	printk(KERN_INFO "joydump: | index: %3d delta: %3d.%02d us data: ", 0, 0, 0);
	for (j = 7; j >= 0; j--)
		printk("%d",(buf[0].data >> j) & 1);
	printk(" |\n");
	for (i = 1; i < t; i++) {
		printk(KERN_INFO "joydump: | index: %3d delta: %3d.%02d us data: ", i,
			(buf[i].time - buf[i-1].time) / div, (buf[i].time - buf[i-1].time) * 100 / div % 100);
		for (j = 7; j >= 0; j--)
			printk("%d",(buf[i].data >> j) & 1);
		printk(" |\n");
	}

	printk(KERN_INFO "joydump: `-------------------- END -------------------'\n");

	return -1;
}

void cleanup_module(void)
{
}


jsattach.c
/*
 * Serial joystick driver attach program.
 *  (c) 1998 David Thompson <dcthomp@mail.utexas.edu>
 *  (c) 1999 Vojtech Pavlik <vojtech@suse.cz>
 *
 * Sponsored by SuSE  
 */

/*
 * This program is used for attaching a serial line to a joystick
 * device, via a line discipline. 
 */

/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * Should you need to contact me, the author, you can do so either by
 * e-mail - mail your message to <vojtech@suse.cz>, or by paper mail:
 * Vojtech Pavlik, Ucitelska 1576, Prague 8, 182 00 Czech Republic
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/time.h>
#include <fcntl.h>
#include <termios.h>
#include <string.h>

struct joy_types {
	char name[16];
	int speed;
	int flags;
	int ldisc;
	int (*init)(int fd);
};

void setline(int fd, int speed, int flags)
{
	struct termios tio;

	tcgetattr(fd, &tio);

	tio.c_cflag = flags | HUPCL | CLOCAL | CREAD;
	tio.c_iflag = IGNBRK | IGNPAR;
	tio.c_oflag = 0;
	tio.c_lflag = 0;

	tio.c_cc[VMIN ] = 1;
	tio.c_cc[VTIME] = 0;
	tio.c_cc[VEOL ] = '\r';
	tio.c_cc[VERASE] = 0;
	tio.c_cc[VKILL] = 0;

	cfsetispeed(&tio, speed);
	cfsetospeed(&tio, speed);

	tcsetattr(fd, TCSANOW, &tio);
}

int readchar(int fd, unsigned char *c, int timeout)
{
	struct timeval tv;
	fd_set set;

	tv.tv_sec = 0;
	tv.tv_usec = timeout * 1000;

	FD_ZERO(&set);
	FD_SET(fd, &set);

	if (!select(fd+1, &set, NULL, NULL, &tv)) return -1;

	read(fd, c, 1);

	return 0;
}

int warrior_init(int fd)
{
	unsigned char c;

	if (readchar(fd, &c, 2000) || c != 'J') return -1;
	while (c != ')' - 0x20) if (readchar(fd, &c, 1000)) return -1;
	if (write(fd, "*S", 2) != 2) return -1;
	if (readchar(fd, &c, 1000) || c != '*' ||
            readchar(fd, &c, 1000) || c != 'S') return -1;
	setline(fd, B4800, CS8);

	return 0;
}

int sball4000_init(int fd)
{
	unsigned char c;

	if (readchar(fd, &c, 4000) || c != 0x11 ||
            readchar(fd, &c, 1000) || c != 0x0d ){
		fprintf(stderr, "This doesn't look like a Spaceball 4000 FLX.\n");
		return -1;
	    }

	sleep(2);			/* Wait a few seconds for the Spaceball to initialize. */

	if (write(fd,"YS\rM\r",5)!=5)	/* Set high-resolution & enable axis events */
		return -1;

	return 0;
}

struct joy_types joy_types[] = {

{ "--sball4",   B9600, CS8,                    12, sball4000_init },
{ "--spaceorb", B9600, CS8,                    15, NULL },
{ "--warrior",  B1200, CS7,                    13, warrior_init },
{ "--magellan", B9600, CS8 | CSTOPB | CRTSCTS, 14, NULL },
{ "", 0, 0 }

};

int main(int argc, char **argv)
{
	int ldisc;
	int type;
	int fd;

	if (argc < 2 || argc > 3 || !strcmp("--help", argv[1])) {
		puts("");
		puts("Usage: jsttach <mode> <device>");
		puts("");
		puts("Modes:");
		puts("  --sball4           SpaceBall 4000 FLX");
		puts("  --spaceorb         SpaceOrb 360 / SpaceBall Avenger");
		puts("  --warrior          WingMan Warrior");
		puts("  --magellan         Magellan / SpaceMouse");
		puts("");
		return 1;
	}
	
	for (type = 0; joy_types[type].speed; type++) {
		if (!strncasecmp(argv[1], joy_types[type].name, 16))
			break;
	}

	if (!joy_types[type].speed) {
		fprintf(stderr, "jsattach: invalid mode\n");
		return 1;
	}

	if ((fd = open(argv[2], O_RDWR|O_NOCTTY|O_NONBLOCK)) < 0) {
		perror("jsattach");
		return 1;
	}

	setline(fd, joy_types[type].speed, joy_types[type].flags);

	if (joy_types[type].init && joy_types[type].init(fd)) {
		fprintf(stderr, "jsattach: device initialization failed\n");
		return 1;
	}

	ldisc = joy_types[type].ldisc;
	if(ioctl(fd, TIOCSETD, &ldisc)) {
		fprintf(stderr, "jsattach: can't set line discipline\n"); 
		return 1;
	}

	while (1) sleep(60);

	ldisc = 0;
	ioctl(fd, TIOCSETD, &ldisc);
	close(fd);

	return 0;
}
